soccer_ipm/soccer_ipm/soccer_ipm.py

```diff
@@ -0,0 +1,323 @@
+        self.declare_parameter('masks.line_mask.scale', 0.0)
+
+        # Parameters
+        self._ball_height = self.get_parameter('ball.ball_radius').value
+        self._bar_height = self.get_parameter('goalposts.bar_height').value
+        self._base_footprint_frame = self.get_parameter('base_footprint_frame').value
+        self._obstacle_footpoint_out_of_image_threshold = \
+            self.get_parameter('obstacles.footpoint_out_of_image_threshold').value
+        self._goalpost_footpoint_out_of_image_threshold = \
+            self.get_parameter('goalposts.footpoint_out_of_image_threshold').value
+        camera_info_topic = self.get_parameter('camera_info.camera_info_topic').value
+        balls_in_image_topic = self.get_parameter('ball.ball_topic').value
+        goalposts_in_image_topic = self.get_parameter('goalposts.goalposts_topic').value
+        obstacles_in_image_topic = self.get_parameter('obstacles.obstacles_topic').value
+        field_boundary_in_image_topic = \
+            self.get_parameter('field_boundary.field_boundary_topic').value
+        line_mask_in_image_topic = self.get_parameter('masks.line_mask.topic').value
+        line_mask_scaling = self.get_parameter('masks.line_mask.scale').value
+
+        # Subscribe to camera info
+        self.create_subscription(CameraInfo, camera_info_topic, self.ipm.set_camera_info, 1)
+
+        self.balls_pub = self.create_publisher(
+            sv3dm.BallArray, 'balls_relative', 1)
+        self.line_mask__pub = self.create_publisher(
+            PointCloud2, 'line_mask_relative_pc', 1)
+        self.goalposts_pub = self.create_publisher(
+            sv3dm.GoalpostArray, 'goal_posts_relative', 1)
+        self.robots__pub = self.create_publisher(
+            sv3dm.RobotArray, 'robots_relative', 1)
+        self.field_boundary_pub = self.create_publisher(
+            sv3dm.FieldBoundary, 'field_boundary_relative', 1)
+
+        # Subscribe to image space data topics
+        self.create_subscription(
+            BallArray, balls_in_image_topic, self.callback_ball, 1)
+        self.create_subscription(
+            GoalpostArray, goalposts_in_image_topic, self.callback_goalposts, 1)
+        self.create_subscription(
+            RobotArray, obstacles_in_image_topic, self.callback_robots, 1)
+        self.create_subscription(
+            FieldBoundary, field_boundary_in_image_topic, self.callback_field_boundary, 1)
+        self.create_subscription(
+            Image,
+            line_mask_in_image_topic,
+            lambda msg: self.callback_masks(
+                msg,
+                self.line_mask__pub,
+                scale=line_mask_scaling), 1)
+
+    def get_field(self, time, heigh_offset=0):
+        plane = PlaneStamped()
+        plane.header.frame_id = self._base_footprint_frame
+        plane.header.stamp = time
+        plane.plane.coef[2] = 1.0  # Normal in z direction
+        plane.plane.coef[3] = -heigh_offset  # Distance above the ground
```

Comment:
If I didn't miss anything `heigh_offset` should be used by e.g. the ball as we project its center point onto an elevated (by its radius) ground plane instead of projecting the footpoint onto the ground, which would be slightly biased if we e.g. look from the top. We can only do this for the ball as it is round and with a known size. This strategy is pretty common.

Reply:
I see, it is being used as you say.